lib/ansible/module_utils/basic.py

```diff
@@ -2621,14 +2622,9 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
             msg = "Argument 'args' to run_command must be list or string"
             self.fail_json(rc=257, cmd=args, msg=msg)
 
-        shell = False
         if use_unsafe_shell:
-            if executable is None:
-                executable = os.environ.get('SHELL')
```

Comment:
not sure why you are removing this, i can see issues with invalid shell/blocker shells set (nologin, /bin/false) but we can work around that.

Reply:
so reread what $SHELL means .. seems i made a mistake, i assumed it was 'current shell' but it is the 'default shell', probably removing these 2 lines will fix most issues, the `else: shell = True` below should stay.

Reply:
Removing these lines will re-break: https://github.com/ansible/ansible/issues/24169

Reply:
OTOH, we might need to find a new fix for #24169 as bcoca is quoting the POSIX definition of SHELL above and the POSIX  definition of popen http://pubs.opengroup.org/onlinepubs/9699919799/  seems to be why this is hardcoded to sh in the first place.

Reply:
not really, this works 'by accident', only when default shell matches the executable requested, as my previous comment states, i misinterpreted what the value of `$SHELL` represents

After rereading, actual fix of the ticket would require we pass ansible_shell_executable information to all modules and fallback to that